### DISCUSSION TITLE
Confidential Containers KBS as a secrets provider

### DISCUSSION THREAD

**Discussion by mythi** [gramineproject/gramine#1334] [Ideas]
Starting with a discussion first to get feedback (and so that I don't forget the idea). 

Currently, Gramine provides a secrets provisioning library that is used by a few "CI examples" to demonstrate how it works. This uses the RA-TLS underneath.

The Confidential Containers (CoCo) project has worked on a generic Key Broker Service with [an attestation protocol for TEEs ](https://github.com/confidential-containers/kbs/blob/main/docs/kbs_attestation_protocol.md). Its original use was to provide container image decryption keys and image pull policies in CoCo but we have also seen demos where unencrypted containers use it to pull data decryption keys.

The idea proposed here is to have a Gramine preload secret provisioning lib that provides secrets from the CoCo KBS. This would add a use-case for an "external" secret provider.

**Comment by kailun-qin** [comment]
> The idea proposed here is to have a Gramine preload secret provisioning lib that provides secrets from the CoCo KBS. This would add a use-case for an "external" secret provider.

@mythi Thanks for initiating the discussion and pls correct me if I'm wrong.

If I understand correctly, you're proposing to add a new "secret provisioning" lib (following the [KBS Attestation Protocol](https://github.com/confidential-containers/kbs/blob/7334c494423fdcca9955fa24b6dd127dcd64f4fa/docs/kbs_attestation_protocol.md) -- basically a protocol over HTTP(S), acting similarly to what KBC does or even at a higher level) in Gramine that can integrate w/ CoCo KBS?

I however see some potential gaps. And I'd rather call the proposed libs w/ a different name (pls see some reasons below).
* First of all, the current Gramine secret provisioning libs (serving as high-level attestation interface) rely heavily on Gramine RA-TLS libs. However, the proposal should be RA-TLS agnostic and may interact directly w/ Gramine low-level `/dev/attestation` interface (e.g., for the `nonce` specified in the protocol). I.e., the designs are quite different.
* Second, the proposal seems to only need provide client-side APIs, whereas Gramine secret prov libs provide APIs for both sides: client (attester) and server (verifer).
* Third, the only APIs that the secret prov libs currently expose are listed here: https://github.com/gramineproject/gramine/blob/4b49ff2ff053857a0e492e5a22aa6d6c107e66d9/tools/sgx/ra-tls/secret_prov.h#L41-L168. And we haven't (yet) considered/expected users to have their own impl. or APIs (see https://github.com/gramineproject/gramine/pull/1114#pullrequestreview-1317146016). If we really want to have it as a Gramine "secret prov" lib, we'd have to evaluate whether these serve well the purpose of the proposal.

Some questions:
1. What APIs or which level of APIs do you expect the proposed libs to expose? (relative to the second and the third point above)
2. Once we have the proposed libs, do we still need a CoCo attestation-agent?
3. Is KBC supplied as libs already? If so, what APIs does it provide, and any plugin mechanisms for attestation evidence? (the reason for this is about whether we need to implement everything from scratch or just an integration w/ KBC)

**Reply by mythi** [reply]
> > The idea proposed here is to have a Gramine preload secret provisioning lib that provides secrets from the CoCo KBS. This would add a use-case for an "external" secret provider.
> 
> @mythi Thanks for initiating the discussion and pls correct me if I'm wrong.
> 
> If I understand correctly, you're proposing to add a new "secret provisioning" lib (following the [KBS Attestation Protocol](https://github.com/confidential-containers/kbs/blob/7334c494423fdcca9955fa24b6dd127dcd64f4fa/docs/kbs_attestation_protocol.md) -- basically a protocol over HTTP(S), acting similarly to what KBC does or even at a higher level) in Gramine that can integrate w/ CoCo KBS?

That's the gist of it, yes.
 
> I however see some potential gaps. And I'd rather call the proposed libs w/ a different name (pls see some reasons below).
> 
>     * First of all, the current Gramine secret provisioning libs (serving as high-level attestation interface) rely heavily on Gramine RA-TLS libs. However, the proposal should be RA-TLS agnostic and may interact directly w/ Gramine low-level `/dev/attestation` interface (e.g., for the `nonce` specified in the protocol). I.e., the designs are quite different.
> 
>     * Second, the proposal seems to only need provide client-side APIs, whereas Gramine secret prov libs provide APIs for both sides: client (attester) and server (verifer).
> 
>     * Third, the only APIs that the secret prov libs currently expose are listed here: https://github.com/gramineproject/gramine/blob/4b49ff2ff053857a0e492e5a22aa6d6c107e66d9/tools/sgx/ra-tls/secret_prov.h#L41-L168
>       . And we haven't (yet) considered/expected users to have their own impl. or APIs (see [[tools/sgx] Install RA-TLS and SecretProv libs properly #1114 (review)](https://github.com/gramineproject/gramine/pull/1114#pullrequestreview-1317146016)). If we really want to have it as a Gramine "secret prov" lib, we'd have to evaluate whether these serve well the purpose of the proposal.

I'm not able to follow what the gaps are in the first two items. My thinking is that this could be an independent "client" shared lib implementation following the KBS API. I do understand that it would have the server side missing from Gramine code base but is that a problem? From CI/validation perspective it probably is which would make it more suitable as a "third party solution", I guess?

> 
> Some questions:
> 
>     1. What APIs or which level of APIs do you expect the proposed libs to expose? (relative to the second and the third point above)

it could piggyback `env.SECRET_PROVISION_*` functionality that I currently feel is sufficient.

>     2. Once we have the proposed libs, do we still need a CoCo attestation-agent?
>
> 
>     3. Is KBC supplied as libs already? If so, what APIs does it provide, and any plugin mechanisms for attestation evidence? (the reason for this is about whether we need to implement everything from scratch or just an integration w/ KBC)

CoCo attestation-agent by default runs a daemon serving gRPC but some of its functions can also be consumed as library dependencies. This provides an abstraction to the underlying KBCes. It's a valid design consideration whether the "secret provider" should use the higher level attestation-agent APIs instead but haven't thought about it too much.

**Comment by mythi** [comment]
the code is in my Trustee repo fork: https://github.com/mythi/trustee/commits/gramine-preload-lib/
